### PR TITLE
[compiler-v2] Add loop labels to the language

### PR DESCRIPTION
**Pull request by wrwg**
## Description

Besides the user being able to describe complex algorithms more efficiently, loop labels are required to express aribtrary reducible control flow graphs in the AST language, and create parity of the AST with the bytecode level for this kind of code (which is also what can be generated from Move).

## How Has This Been Tested?

New baseline tests in checking, v1-checking, bytecode generation, and a transactional test.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [x] Move Compiler
- [ ] Other (specify)

### CODE REVIEW

**Review by vineethk — COMMENTED** [review]
Looking pretty good! I have a few comments, but should be ready to go once they are addressed.

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 1]
```diff
@@ -1167,17 +1167,20 @@ fn parse_term(context: &mut Context) -> Result<Exp, Box<Diagnostic>> {
...
                 return Err(diag);
             }
-            Exp_::Break
+            Exp_::Break(label)
         },
 
         Tok::Continue => {
             context.tokens.advance()?;
-            Exp_::Continue
+            let label = parse_optional_label(context)?;
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:1182]: Continue with a value wouldn't make sense, I think that is why its not dealt with.

File: third_party/move/move-compiler/src/parser/syntax.rs [hunk 2]
```diff
@@ -1392,6 +1395,17 @@ fn parse_spec_loop_invariant(context: &mut Context) -> Result<SequenceItem, Box<
 // should be,     if (cond) e1 else (e2 + 1)
 fn parse_control_exp(context: &mut Context) -> Result<(Exp, bool), Box<Diagnostic>> {
     let start_loc = context.tokens.start_loc();
+    let label = parse_optional_label(context)?;
+    if label.is_some() {
+        consume_token(context.tokens, Tok::Colon)?;
+        // Check here whether a label is actually allowed
+        if !matches!(context.tokens.peek(), Tok::While | Tok::Loop) {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/syntax.rs:1402]: The for is not implemented here but as syntactic sugar elsewhere in the parser. Its a special case which we resolve in the parser. When its refactored and generalized (for becomes an AST node) we can do this.

**Review by wrwg — COMMENTED** [review]
Thanks for the review!

**Review by vineethk — APPROVED** [review]
Looks good, have one comment to be addressed.

File: third_party/move/move-compiler/src/parser/lexer.rs [hunk 3]
```diff
@@ -505,6 +507,10 @@ fn find_token(
                 (get_name_token(&text[..len]), len)
             }
         },
+        '\'' => {
+            let len = get_name_len(&text[1..]);
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler/src/parser/lexer.rs:520]: Good find. This is a problem in the lexer. Fixed and added some test cases.

**Review by wrwg — COMMENTED** [review]
Comments addressed, thanks for the review!

**Review by brmataptos — APPROVED** [review]

**Pull request merged**